HW-17: Добавлено логирование исходящих запросов

DrEremin opened the pull request:
Добавлены классы:
    RequestData, ResponseData, RequestDirection, FeignLogger, LoggingService;
Добавлен метод RequestDataFormatter.formatFeignRequestHeaders(); 
В логику классов WebLoggingFilter и WebLoggingRequestBodyAdvice добавлен
    LoggingService + удалены лишние зависимости;
В конфигурационный класс LoggingStarterAutoConfiguration добавлены новые бины; 
Добавлена зависимость для клиента Feign в build.gradle.

DrEremin committed:
HW-17: Добавлено логирование исходящих запросов

Добавлены классы:
    RequestData, ResponseData, RequestDirection, FeignLogger, LoggingService;
Добавлен метод RequestDataFormatter.formatFeignRequestHeaders;
В логику классов WebLoggingFilter и WebLoggingRequestBodyAdvice добавлен
    LoggingService, удалены лишние зависимости;
В конфигурационный класс LoggingStarterAutoConfiguration добавлены новые бины;
Добавлена зависимость для клиента Feign в build.gradle.

PanyukovNN commented on src/main/java/ru/dreremin/loggingstarter/dto/RequestData.java:
```diff
@@ -0,0 +1,85 @@
+    private String headers;
+    private String body;
+
+    public RequestData() {
```
для чего нам этот конструктор?

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/dto/RequestData.java:
```diff
@@ -0,0 +1,85 @@
+        return new Builder(new RequestData());
+    }
+
+    public static class Builder {
```
Обычно билдер идет в самом низу класса, после всех методов

DrEremin replied:
Переставил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/dto/RequestData.java:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        public Builder method(String method) {
+            requestData.setMethod(method);
```
это неправильна реализация билдера, он не проставляет значения через setter, а у него должен быть тот же набор полей что и у сущности, которую он создает

сначала проставляются значения в эти поля внутри билдера, затем только создается сама сущность в момент вызова метода build, через отдельный конструктор

DrEremin replied:
Билдеры исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/dto/RequestData.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    public void setDirection(RequestDirection direction) {
+        this.direction = direction.name();
```
в сеттерах вообще лучше никакой логики не держать, здесь у тебя идет вызов .name() и он может неожиданно выбросить NPE

DrEremin replied:
Сеттеры убрал при исправлении билдеров

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/RequestDataFormatter.java:
```diff
@@ -23,6 +22,16 @@ public static String formatRequestHeaders(HttpServletRequest request, Set<String
+    public static String formatFeignRequestHeaders(Request request, Set<String> excludedHeaders) {
+        Map<String, Collection<String>> headers = request.headers();
+        String inlineHeaders = headers.keySet().stream()
+                .filter(it -> !excludedHeaders.contains(it))
```
важно выполнять проверку с игнорированием регистра

DrEremin replied:
Сделал. Правда пришлось и в классе свойств их приводить к нижнему регистру в postConstruct-методе, чтобы не делать это каждый раз.

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/service/LoggingService.java:
```diff
@@ -0,0 +1,96 @@
+                .body(prepareBody(body))
+                .build();
+
+        log.info("Запрос: {}", requestData);
```
Тут надо написать "Тело запроса: ..."

и не обязательно в логи выводить заголовки

DrEremin replied:
Оставил пока, потому при запросе с телом, весь запрос выводится в одном логе (на мой взгляд так лучше читается), иначе будет запрос и тело  в разных логах. Если все же нужно сделать раздельными, то исправлю.

panyukovnn replied:
В целом вариант неплохой, только я не уверен что в 100% случаев будет корректно работать проверка request.getContentLength() == -1 в вебфильтре, но пусть пока так остается

DrEremin committed:
HW-17: небольшие исправления

Исправлена реализация паттерна Builder в классах RequestData и ResponseData;
Добавлена проверка заголовков на исключение из логирования без учета регистра.

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/property/LoggingStarterProperties.java:
```diff
@@ -15,11 +13,23 @@ public class LoggingStarterProperties {
         this.uriPaths = new ArrayList<>();
     }
 
+    @PostConstruct
```
Плохая практика, так делать не стоит, лучше для сравнения с игнорированием регистра использовать готовые инструменты, например есть метод equalsIgnoreCase в StringUtils библиотеки commons-lang3

DrEremin replied:
Исправил

DrEremin committed:
HW-17: небольшие исправления

Поле LoggingStarterProperties.headers было Set стало List;
Удален init-метод класса LoggingStarterProperties;
Поиск заголовка среди исключаемых заголовков осуществлен перебором по стриму.

DrEremin committed:
HW-17: обновлены исполяемые файлы библиотеки до версии 1.8-SNAPSHOT
